client/components/AdminPanel.jsx

```diff
@@ -0,0 +1,165 @@
+            isCreatingNewUser: false
+        });
+
+        this.closeModal();
+    }
+
+    async submitUpdateUser(user) {
+        const { email, password, isAdmin } = user;
+        const userId = this.state.selectedUserToUpdate._id;
+        await this.props.updateUser(userId, email, password, isAdmin);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+
+        this.closeModal();
+    }
+
+    closeModal() {
+        this.setState({
+            showModal: false,
+            selectedUserToUpdate: {}
+        });
+    }
+
+    async deleteUser(user) {
+        await this.props.deleteUser(user._id);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+    }
+
+    render() {
+        if (!this.props.users) {
+            return <div></div>;
+        }
+
+        const { totalPages, totalEntries, currentPage, users } = this.props;
+
+        return (
+            <div>
+                <header className="navbar my-2">
+                    <section className="navbar-section">
+                        <div className="input-group input-inline">
+                            <input 
+                                className="form-input" type="text" placeholder="Filter users" 
+                                value={this.state.filterString}
+                                onChange={event => this.setState({ filterString: event.target.value })}
+                            />
+                            <button className={`btn btn-primary input-group-btn ${this.state.showModal ? 'disabled' : ''}`} onClick={() => this.fetchPage(1)}>Filter</button>
+                        </div>
+                    </section>
+                    <section className="navbar-section">
+                        <button 
+                            className="btn"
+                            onClick={() => this.setState({ showModal: !this.state.showModal, isCreatingNewUser: true })}
+                        >
+                            + Add User
+                        </button>
+                    </section>
+                </header>
+                {users.map(user => {
+                    return (
+                        <UserTile key={user._id} user={user} onUpdateUser={this.openUpdateModal} onDeleteUser={this.deleteUser} /> 
+                    );
+                })}
+                <div className="column col-6 col-mx-auto">
+                    <Pagination numPages={totalPages} totalEntries={totalEntries} currentPage={currentPage} onPageClick={this.fetchPage} />                
+                </div>
+                {this.state.showModal &&
+                    <UserModal onSubmitUser={this.onSubmitModal} selectedUser={this.state.selectedUserToUpdate} close={this.closeModal} />
+                }
+                {this.props.error &&
+                    <div className="toast toast-error">
+                        {this.props.message}
+                    </div>
+                }
```

Comment:
all of the potential errors are wrapped up in one object, so if there is an error of any kind, we render it here